knowledge-base/implementation-documentation/api-improvements.md

```diff
@@ -0,0 +1,209 @@
+* The examples below could be non-breaking changes done sequentially over time.
+
+Futher examples (and options):
+```ts
+BusinessPartner.requestBuilder().getAll()
+  .filter((bupa, { and }) => and(
+    bupa.FIRST_NAME.equals('Peter')
+  ))
+```
+
+```ts
+BusinessPartner.requestBuilder().getAll()
+  .filter((bupa, { and, equals }) => and(
+    equals(bupa.FIRST_NAME, 'Peter')
+  ))
+```
+
+Using a JavaScript function:
+```ts
+BusinessPartner.requestBuilder().getAll()
+  .filter(buPa => buPa.FIRST_NAME === 'Peter' && buPa.LAST_NAME === 'Pan')
+```
+* Difficult to implement.
+* Not realistic for 2.0, but maybe later.
+
+## Get rid of `execute`
+
+### Current API
+```ts
+const buPa: BusinessPartner = await BusinessPartner.requestBuilder().getAll().execute(destination);
+```
+
+### Issues
+
+### Alternatives
+
+```ts
+const buPa: BusinessPartner = await BusinessPartnerApi(destination).getAll({
+  filter: ...,
+  select: ...,
+  top: ...
+});
+```
+* Not realistic for 2.0, but maybe later.
+
+## Batch
+
+### Current API
+
+```ts
+// type BatchResponse = WriteResponses | ReadResponse | ErrorResponse
+const responses: BatchResponse[] = await batch( // The name batch response implies that this is the whole response of a batch, while it is a subresponse here
+  getAllRequestBuilder,
+  changeset(
+    createRequestBuilder,
+    deleteRequestBuilder,
+    updateRequestBuilder
+  )
+).execute(destination);
+
+if (responses.every(response => response.isSucces())) { // Why is isSuccess a function?
+  const bupas = (responses[0] as ReadResponse).as(BusinessPartner) // It is unfortunate that I have to cast
+  const changeSetResponse: WriteResponses = responses[1] as WriteResponses;
+  if (changeSetResponse.responses[0].httpCode !== 204) {
+    const createResponse = changeSetResponse.responses[0].as!(BusinessPartner);
+  }
+  const deleteResponse = changeSetResponse.responses[1].as!(BusinessPartner);
+}
+```
+
+### Issues
+
+### Alternatives
+Quick wins:
+* Change isSuccess() to isSuccess
+* Rename WriteResponses to ChangesetResponse
+
+Investigation needed:
+* Root level types with 10 arguements (generics) for batch and changeset
+* If possible map incoming requests to responses (x-request-id, content-id, message-id headers?)
+
+
+## Execute raw
+
+### Current API
+```ts
+const buPa: BusinessPartner = await BusinessPartner.requestBuilder().getAll().execute(destination);
+```
+
+### Issues
+
+### Alternatives
+
+```ts
+const buPa: Response<BusinessPartner> = await BusinessPartner.requestBuilder().getAll().executeRaw(destination);
+```
+
+```ts
+const [buPa, req, res]: [BusinessPartner, Request, Response] = await BusinessPartner.requestBuilder().getAll().execute(destination);
+```
+
+```ts
+const [buPa, req, res]: [BusinessPartner, Request, Response] = await BusinessPartner.requestBuilder().getAll().executeRaw(destination);
+const buPa: BusinessPartner = await BusinessPartner.requestBuilder().getAll().execute(destination);
+```
```

Comment:
```suggestion
```
I (FE) find this options most appealing since it will allow for extra information but in the cases the users do not like it they do not have do use the deconstruction. We can potentially also deprecate the plain `execute` and push users in the direction of the `executeRaw` method. BTW the name `executeRaw` is of course not fitting to the return type `[buPa, req, res]`.  What do you think about a `showRequest` and `showResponse` method which will change the return type form a `BusinessPartner` to a [BusinessPartner,Res,Req]?
```